content/docs/command-reference/exp/run.md

```diff
@@ -0,0 +1,141 @@
+# exp run
+
+Run an experiment: reproduce a variation of a committed
+[pipeline](/doc/command-reference/dag) in a hidden project branch.
+
+> Similar to `dvc repro` but for
+> [experimentation](/doc/user-guide/experiment-management).
+
+## Synopsis
+
+```usage
+usage: dvc exp run [-h] [-q | -v] [-f]
+                   [<repro_options> ...]
+                   [--params [<filename>:]<params_list>]
+                   [-n <name>] [--queue] [--run-all] [-j <number>]
+                   [targets [targets ...]]
+
+positional arguments:
+  targets               Stages to reproduce. 'dvc.yaml' by default.
+```
+
+## Description
+
+Provides a way to execute and track `dvc experiments` in your
+<abbr>project</abbr> without polluting it with unnecessary commits, branches,
+directories, etc.
+
+> `dvc exp run` is equivalent to `dvc repro` for <abbr>experiments</abbr>. It
+> has the same behavior when it comes to `targets` and stage execution (restores
+> the dependency graph, etc.). See the command [options](#options) for more the
+> differences.
+
+Before using this command, you'll probably want to make modifications such as
+data and code updates, or <abbr>hyperparameter</abbr> tuning. You can use the
+`--set-param` option to change `dvc param` values on-the fly.
+
+Each `dvc exp run` creates a variation based on the latest project version
+committed to Git, and tracks this experiment automatically with an automatic
+name like `exp-bfe64` (which can be customized with the `--name` option).
+
+<details>
+
+### How does DVC track experiments?
+
+Internally, `dvc exp` uses actual commits under custom
+[Git references](https://git-scm.com/book/en/v2/Git-Internals-Git-References)
+(found in `.git/refs/exps`). Each commit has the Git `HEAD` as parent and has
+it's own SHA-256 hash. These are not pushed to the Git remote by default (see
+`dvc exp push`).
+
+> References have a unique signature similar to the
+> [entries in the run-cache](/doc/user-guide/project-structure/internal-files#run-cache).
+
+</details>
+
+The results of the last experiment can be seen in the <abbr>workspace</abbr>. To
+display and compare your experiments, use `dvc exp show` or `dvc exp diff`. Use
+`dvc exp apply` to roll back the workspace to a previous experiment
+
+Successful experiments can be made
+[persistent](/doc/user-guide/experiment-management#persistent-experiments) by
+committing them to the Git repo. Unnecessary ones can be removed with
+`dvc exp gc`, or abandoned (and their data with `dvc gc`).
+
+## Checkpoints
+
+To track successive steps in a longer <abbr>experiment</abbr>, you can register
+checkpoints with DVC during your code or script runtime (similar to a logger).
+
+To do so, mark stage `outs` with `checkpoint: true` in `dvc.yaml`, and in your
+code either call the `dvc.api.make_checkpoint()` function (Python), or write a
```

Comment:
> mark stage `outs` with `checkpoint: true` in `dvc.yaml`, and...

@pmrowla so is this required for checkpoints to work or is it just optional? Other than stage repro/checkoiut not failing when a checkpoint output doesn't exist, what are the differences? I'm trying to understand what this is for, still a bit unclear... Thanks

Reply:
It is required for checkpoints to work, if a stage does not contain any `checkpoint` outputs, it will be reproduced like a regular standalone stage, and DVC will not monitor for the `make_checkpoint()` signal file (and will not create any checkpoint commits).

Reply:
One other difference is that `checkpoint` outs are also treated similarly to `persist` outs, and are not removed when reproducing a stage. The difference between `checkpoint` and `persist` outs is that a `checkpoint` out will be reverted to its last committed state at the start of reproducing a stage, and DVC will not touch a regular `persist` out at all at the start of reproducing a stage.

Reply:
Basically checkpoint outs are circular dependencies, where a truly reproducible stage should have independently tracked input (DVC dependency ) and output (DVC out) versions of the single checkpoint file. In DVC we don't have proper internal support for circular dependencies, but the `checkpoint` out behavior is a good enough workaround for now.

Reply:
Thanks for all the details as usual @pmrowla, I have mentioned them in docs now.

> required for checkpoints to work

Just wondering why we want/need this restriction. Couldn't DVC always monitor that signal file regardless? Perhaps it's a significant overhead to do so?

> a checkpoint out will be reverted to its last committed state at the start of reproducing a stage

But only with `exp run` right? Not with `repro`

Reply:
Also should it be `checkpoints: true`? (plural)

Reply:
`dvc repro` will abort if a stage contains any `checkpoint` outs, checkpoint outs explicitly are only compatible with `dvc exp run`.

I also don't see why it should be plural, since it is a flag for a singular out? You can have a stage that contains more than one checkpoint out, the same way that you can have a stage containing more than one uncached/persist/etc out, but the flag still only applies per individual out

Reply:
> Just wondering why we want/need this restriction. Couldn't DVC always monitor that signal file regardless? Perhaps it's a significant overhead to do so?

There is nothing preventing us from doing this, but it was written this way given that this is new behavior, and it's not exactly clear what DVC should be doing for stages that don't contain explicit checkpoint outs. For checkpoint outs, the user's stage should still be reproducible in between checkpoints (given an input checkpoint out state, the next output should always be the same). For "regular" stages, the expectation is that only the full stage is reproducible, rather than arbitrary smaller parts within the stage (so DVC should not create commits for any intermediate state within the stage).

Reply:
> dvc repro will abort if a stage contains any checkpoint outs

Just noticed this answer. Hm, interesting... OK I'll document all this and we can move the discussion if any to the issue I created in core. Thanks!